k2/csrc/array_ops_benchmark.cu

```diff
@@ -0,0 +1,98 @@
+/**
+ * @brief Benchmarks for k2 APIs.
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include "k2/csrc/array_ops.h"
+#include "k2/csrc/log.h"
+#include "k2/csrc/timer.h"
+
+namespace k2 {
+
+// ------- from google/benchmark  ---- begin ----
+#if defined(__GNUC__)
+#define K2_ALWAYS_INLINE __attribute__((always_inline))
+#else
+#define K2_ALWAYS_INLINE
+#endif
+
+// The DoNotOptimize(...) function can be used to prevent a value or
+// expression from being optimized away by the compiler. This function is
+// intended to add little to no overhead.
+// See: https://youtu.be/nXaxk27zwlk?t=2441
+template <class Tp>
+inline K2_ALWAYS_INLINE void DoNotOptimize(Tp const &value) {
+  asm volatile("" : : "r,m"(value) : "memory");
+}
+
+template <class Tp>
+inline K2_ALWAYS_INLINE void DoNotOptimize(Tp &value) {
+#if defined(__clang__)
+  asm volatile("" : "+r,m"(value) : : "memory");
+#else
+  asm volatile("" : "+m,r"(value) : : "memory");
+#endif
+}
+// ------- from google/benchmark  ---- end ----
+
+/* Return the number of milliseconds per iteration on average.
+
+   @param  [in]  dim  Number of elements in Array1<T> for testing.
+   @return Return number of ms per iteration on avarage.
+ */
+
+template <typename T>
+using UnaryOp = Array1<T> (*)(const Array1<T> &);
```

Comment:
We can make it a variadic template that accepts an arbitrary number of arguments.

Reply:
Bear in mind that not all the functions we might want to call output their data via return values.
I'm not objecting, just pointing it out.